File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
'''
1.아이디어
    - item을 하나씩 돌면서, startswith()를 해서 확인한다 하나라도 있으면 true 없으면 false
        -> n(n-1)번 반복해야 하므로, 시간복잡도 매우 초과 -> set()으로 바꿔서 찾으면 더 빨라진다
2.시간복잡도
    - list로도 가능한 연산이지만, 매우 많은 길이가 주어질 때 set 해시구조를 통해 복잡도를 줄일 수 있다.
'''
def solution(phone_book):
    phone_set = set(phone_book)
    
    for number in phone_book:
        prefix = ""
        for char in number[:-1]:
            prefix += char
            if prefix in phone_set:
                return False
    return True 

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
/*
[아이디어]
1. 각각의 전화번호를 오름차순 정렬 - ONlogN)
2.startsWith()로 시작하는 게 있는지 확인하기 - O(N)
4. 있으면 false 반환, 없으면 true - O(1)
*/
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
       
        Arrays.sort(phone_book, (a,b) -> {
            return a.compareTo(b);
        });
        
        for(int i=0; i<phone_book.length-1; i++){
            if(phone_book[i+1].startsWith(phone_book[i])) return false;
        }
        
        return true;
    }
}